### PR TITLE
Allow requesting IRCv3 extended-join capability

### PR DESCRIPTION
**Pull request by Fusl**
Standard IRC `JOIN` messages look like this:
```
:nick!ident@host JOIN #channelname
```

IRCv3 `JOIN` messages look like this:
```
:nick!ident@host JOIN #channelname * :Fusl
```
or
```
:nick!ident@host JOIN #channelname authenticated-username :Fusl
```

`message.args[1]` will be the authenticated username and `message.args[2]` the real name.

### CODE REVIEW

File: lib/irc.js [hunk 1]
```diff
@@ -633,6 +634,13 @@ function Client(server, clientNick, opt) {
                 // currently only handle ACK sasl responses
```
Comment (Throne3d) [inline lib/irc.js]: Nitpick: this no longer only handles ACKs for sasl, so could you amend this? ("currently only handles ACK sasl and extended-join", for example)

File: lib/irc.js [hunk 2]
```diff
@@ -752,6 +760,10 @@ Client.prototype._connectionHandler = function() {
     if (this.opt.webirc.ip && this.opt.webirc.pass && this.opt.webirc.host) {
         this.send('WEBIRC', this.opt.webirc.pass, this.opt.userName, this.opt.webirc.host, this.opt.webirc.ip);
     }
+    
+    if (this.opt.extendedJoin) {
```
Comment (Throne3d) [inline lib/irc.js]: Can you add a link to https://ircv3.net/specs/extensions/extended-join-3.1.html here?

File: lib/irc.js [hunk 3]
```diff
@@ -68,7 +68,8 @@ function Client(server, clientNick, opt) {
         encoding: null,
         millisecondsOfSilenceBeforePingSent: 15 * 1000,
         millisecondsBeforePingTimeout: 8 * 1000,
-        enableStrictParse: false
+        enableStrictParse: false,
+        extendedJoin: false
```
Comment (Throne3d) [inline lib/irc.js]: If you could add a note to the API documentation about `extendedJoin`, that would be great.

**Comment by Throne3d** [comment]
I'm a little confused about the motivation behind this PR - without [account-notify](https://ircv3.net/specs/extensions/account-notify-3.1.html), won't [extended-join](https://ircv3.net/specs/extensions/extended-join-3.1.html) do very little to modify the experience as a user of the library? I'm okay accepting it anyway, it just seems like it won't be very useful.

I'd like this code to interact well with a `CAP NAK` response, so that if we request `extended-join`, and get told that it's not supported, we still successfully respond with `CAP END`. I'm happy if this is emitted as an error at the same time.

Could you also add some tests for this? Primarily, ensure that:
- it requests the capability when enabled and SASL is disabled, and responds with `CAP END` if it receives a response
- it does not request the capability when disabled
- with SASL enabled, it requests both `sasl` and `extended-join`, then waits for a SASL-related CAP response before sending `CAP END`

(I'm okay with, for now, assuming the responses come in a timely and ordered fashion, and accordingly that we won't enter a state where we have a response for `sasl` but not for `extended-join` for a prolonged amount of time - this could cause weirdness where we respond with `CAP END` and `extended-join` fails to get registered, but it seems unlikely to happen. Ideally, if any future capability negotiation is added to this library, the system will be refactored.)

Thanks for your contribution!

**Review by Throne3d — CHANGES_REQUESTED** [review]
(Per comment)

**Comment by Fusl** [comment]
> I'm a little confused about the motivation behind this PR - without account-notify, won't extended-join do very little to modify the experience as a user of the library?

I have a bot that automatically voices users and voices them immediately when they are authenticated: https://github.com/opennic/irc-bot-openvoice/blob/master/openvoice.js#L121

**Comment by tritium21** [comment]
@Throne3d The information that is already on the bot config or the welcome message is the *bot*'s authenticated name, etc.  The extended join message also adds authenticated names for all *other users* who join a channel.

**Comment by Throne3d** [comment]
Any chance you'll get to the requested changes soon?

**Pull request closed without merging**